BSP/STM32F413/Src/BSP_CAN.c

```diff
@@ -0,0 +1,402 @@
+    CAN_InitStruct.CAN_ABOM = DISABLE;
+    CAN_InitStruct.CAN_AWUM = DISABLE;
+    CAN_InitStruct.CAN_NART = DISABLE;
+    CAN_InitStruct.CAN_RFLM = DISABLE;
+    CAN_InitStruct.CAN_TXFP = DISABLE;
+    CAN_InitStruct.CAN_Mode = CAN_Mode_LoopBack; // TODO: change back to Normal after testing
+    CAN_InitStruct.CAN_SJW = CAN_SJW_1tq;
+
+    /* CAN Baudrate = 125 KBps
+        * 1/(prescalar + (prescalar*BS1) + (prescalar*BS2)) * Clk = CAN Baudrate
+        * The CAN clk is currently set to 20MHz (APB1 clock set to 20MHz in BSP_PLL_Init())
+    */
+    CAN_InitStruct.CAN_BS1 = CAN_BS1_3tq;
+    CAN_InitStruct.CAN_BS2 = CAN_BS2_4tq;
+    CAN_InitStruct.CAN_Prescaler = 16;
+    CAN_Init(CAN3, &CAN_InitStruct);
+
+    /* CAN filter init */
+    CAN_FilterInitStruct.CAN_FilterNumber = 0;
+    CAN_FilterInitStruct.CAN_FilterMode = CAN_FilterMode_IdMask;
+    CAN_FilterInitStruct.CAN_FilterScale = CAN_FilterScale_32bit;
+    CAN_FilterInitStruct.CAN_FilterIdHigh = 0x0000;
+    CAN_FilterInitStruct.CAN_FilterIdLow = 0x0000;
+    CAN_FilterInitStruct.CAN_FilterMaskIdHigh = 0x0000;
+    CAN_FilterInitStruct.CAN_FilterMaskIdLow = 0x0000;
+    CAN_FilterInitStruct.CAN_FilterFIFOAssignment = 0;
+    CAN_FilterInitStruct.CAN_FilterActivation = ENABLE;
+    CAN_FilterInit(CAN3, &CAN_FilterInitStruct);
+
+    //CAN_SlaveStartBank(CAN1, 0);
+
+    /* Transmit Structure preparation */
+    gTxMessage[1].ExtId = 0x5;
+    gTxMessage[1].RTR = CAN_RTR_DATA;
+    gTxMessage[1].IDE = CAN_ID_STD;
+    gTxMessage[1].DLC = 1;
+
+    /* Receive Structure preparation */
+    gRxMessage[1].StdId = 0x00;
+    gRxMessage[1].ExtId = 0x00;
+    gRxMessage[1].IDE = CAN_ID_STD;
+    gRxMessage[1].DLC = 0;
+    gRxMessage[1].FMI = 0;
+
+    /* Enable FIFO 0 message pending Interrupt */
+    CAN_ITConfig(CAN3, CAN_IT_FMP0, ENABLE);
+
+    // Enable Rx interrupts
+    NVIC_InitStruct.NVIC_IRQChannel = CAN3_RX0_IRQn;
+    NVIC_InitStruct.NVIC_IRQChannelPreemptionPriority = 0x00;
+    NVIC_InitStruct.NVIC_IRQChannelSubPriority = 0x00;
+    NVIC_InitStruct.NVIC_IRQChannelCmd = ENABLE;
+    NVIC_Init(&NVIC_InitStruct);	
+
+}
+
+/**
+ * @brief   Transmits the data onto the CAN bus with the specified id
+ * @param   id : Message of ID. Also indicates the priority of message. The lower the value, the higher the priority.
+ * @param   data : data to be transmitted. The max is 8 bytes.
+ * @param   length : num of bytes of data to be transmitted. This must be <= 8 bytes or else the rest of the message is dropped.
+ * @return  ERROR if module was unable to transmit the data onto the CAN bus. SUCCESS indicates data was transmitted.
+ */
+ErrorStatus BSP_CAN_Write(CAN_t bus, uint32_t id, uint8_t data[8], uint8_t length) {
+    
+    gTxMessage[bus].StdId = id;
+    gTxMessage[bus].DLC = length;
+	for(int i = 0; i < length; i++){
+        gTxMessage[bus].Data[i] = data[i];
+    }
+
+    uint8_t retVal = (CAN_Transmit(bus == CAN_1 ? CAN1 : CAN3, &gTxMessage[bus]) != 0);
```

Comment:
FYI, `CAN_Transmit()` returns the mailbox it used, so you are returning errors, when you shouldn't (if it uses mailbox 1 or 2). I would refer to [BPS's implementation](https://github.com/lhr-solar/BPS/blob/6eb5140300e02cd123a3e009d797d15ab42d0769/BSP/STM32F413/Src/BSP_CAN.c#L156) to fix this